// chat-engine-unread-messages/dist/chat-engine-unread-messages.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
(function() {

    const package = require('../package.json');
    window.OpenChatFramework.plugin[package.name] = require('../src/plugin.js');

})();

},{"../package.json":2,"../src/plugin.js":3}],2:[function(require,module,exports){
module.exports={
  "author": "Ian Jennings",
  "name": "chat-engine-unread-messages",
  "version": "0.0.2",
  "main": "src/plugin.js",
  "dependencies": {
    "chat-engine": "^0.0.4"
  }
}

},{}],3:[function(require,module,exports){
module.exports = (config) => {

    class extension {

        construct(data) {

            this.parent.isFocused = false;
            this.parent.unreadCount = 0;

            this.parent.on('message', (event) => {

                if(!this.isActive) {

                    this.parent.unreadCount++;
                    this.parent.trigger('$unread', {
                        chat: this.parent,
                        sender: event.sender,
                        event: event
                    });

                }

            });

        }

        active() {

            this.isActive = true;
            this.parent.unreadCount = 0;
        }

        inactive() {
            this.isActive = false;
        }

    };

    // attach methods to Chat
    return {
        namespace: 'unreadMessages',
        extends: {
            Chat: extension
        }
    }

}

},{}]},{},[1])
