// ocf-history/src/plugin.js
// define our plugin values in root
const defaults = {timeout: 1000};

module.exports = (config) => {

    class extension {

        // construct is run automatically when plugin is added to Class
        construct(config = {}) {

            config.channel = config.channel || this.parent.channel;

            this.OCF.pubnub.history(config, (status, response) => {

                if(response.error) {
                    throw new Error(response.error);
                } else {

                    response.messages.forEach((message) => {

                        // trigger the same event with the same data
                        // but the event name is now history:name rather than just name
                        // to distinguish it from the original live events
                        this.parent.trigger(
                            ['$history', message.entry[0]].join('.'),
                            message.entry[1]);

                    });

                }

            });

        }
    };

    // attach methods to Chat
    return {
        namespace: 'history',
        extends: {
            Chat: extension,
            GlobalChat: extension
        }
    }

}
