// ocf-history/dist/ocf-history.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
(function() {

    const package = require('../package.json');
    window.OpenChatFramework.plugin[package.name] = require('../src/plugin.js');

})();

},{"../package.json":2,"../src/plugin.js":3}],2:[function(require,module,exports){
module.exports={
  "name": "ocf-history",
  "version": "0.0.1",
  "main": "src/plugin.js",
  "dependencies": {
    "ocf": "^0.0.4"
  },
  "devDependencies": {
    "chai": "^3.5.0",
    "mocha": "^3.3.0"
  }
}

},{}],3:[function(require,module,exports){
// define our plugin values in root
const defaults = {timeout: 1000};

module.exports = (config) => {

    class extension {

        // construct is run automatically when plugin is added to Class
        construct(data) {

            // we can access the rltm room connection and call it's history method
            this.parent.room.history().then((response) => {

                response.reverse();

                // for every message we get back
                for(let i in response) {


                    if(response[i].data) {

                        // broadcast the same event with the same data
                        // but the event name is now history:name rather than just name
                        // to distinguish it from the original live events
                        this.parent.broadcast(
                            ['$history', response[i].data.message[0]].join('.'),
                            response[i].data.message[1]);

                    } else {

                        // something went wrong, person probably doesn't have history enabled
                        throw new Error(response[i]);

                    }

                }

            }, (error) => {
                throw new Error('There was a problem getting message history', error);
            });

        }
    };

    // attach methods to Chat
    return {
        namespace: 'history',
        extends: {
            Chat: extension,
            GlobalChat: extension
        }
    }

}

},{}]},{},[1])
